File: internal/config/monitor.go
package config

import (
	"time"
)

// ServiceConfig 单个服务监测配置
type ServiceConfig struct {
	Provider     string       `yaml:"provider" json:"provider"`
	ProviderName string       `yaml:"provider_name" json:"provider_name,omitempty"` // Provider 显示名称（可选，未配置时回退到 provider）
	ProviderSlug string       `yaml:"provider_slug,omitempty" json:"provider_slug"` // URL slug（可选，未配置时使用 provider 小写）
	ProviderURL  string       `yaml:"provider_url,omitempty" json:"provider_url"`   // 服务商官网链接（可选）
	Service      string       `yaml:"service" json:"service"`
	ServiceName  string       `yaml:"service_name" json:"service_name,omitempty"`   // Service 显示名称（可选，未配置时回退到 service）
	Category     string       `yaml:"category" json:"category"`                     // 分类：commercial（商业站）或 public（公益站）
	Sponsor      string       `yaml:"sponsor,omitempty" json:"sponsor"`             // 赞助者：提供 API Key 的个人或组织
	SponsorURL   string       `yaml:"sponsor_url,omitempty" json:"sponsor_url"`     // 赞助者链接（可选）
	SponsorLevel SponsorLevel `yaml:"sponsor_level,omitempty" json:"sponsor_level"` // 赞助等级：public/signal/pulse/beacon/backbone/core（可选，按通道赞助）
	KeyType      string       `yaml:"key_type" json:"key_type,omitempty"`           // API Key 类型：official/user（空值按 official 处理）
	PriceMin     *float64     `yaml:"price_min,omitempty" json:"price_min"`         // 参考倍率下限（可选，如 0.05）
	PriceMax     *float64     `yaml:"price_max,omitempty" json:"price_max"`         // 参考倍率（可选，如 0.2）
	Annotations  []Annotation `yaml:"-" json:"annotations,omitempty"`               // 统一注解（系统派生 + annotation_rules）
	Channel      string       `yaml:"channel" json:"channel"`                       // 业务通道标识（如 "vip-channel"），用于分类和过滤
	Model        string       `yaml:"model" json:"model,omitempty"`                 // 模型系列名（展示/DB 键；可由 template 提供，config 可覆盖）
	ModelID      string       `yaml:"model_id,omitempty" json:"model_id,omitempty"` // 监测行稳定唯一 id（md_<uuidv4>），系统生成、不可变；与 Model 一样不参与父子继承。Phase 1 起作内部存储键
	ChannelID    string       `yaml:"-" json:"-"`                                   // 运行时注入：所属文件 channel_id 下沉到行（loadMonitorsDir），不持久化，经 MonitorResult.channel_id 暴露
	RequestModel string       `yaml:"request_model,omitempty" json:"-"`             // 实际请求模型 ID（注入 {{MODEL}}/{{REQUEST_MODEL}}，为空时回退 Model）
	// ModelVendor 模型厂商受控 code（词表见 internal/modelvendor）。与 Model/RequestModel 一样
	// 可由 template 提供、config 行级覆盖；与 RequestModel 一样参与父子继承——同一通道的各 model
	// 层必属同一厂商（「一个通道一个厂商」不变量，见 validateModelVendors）。
	// 注意它与 Model 的继承语义相反：Model 是父子的区分字段故不继承，vendor 是通道内共享字段。
	ModelVendor string `yaml:"model_vendor,omitempty" json:"model_vendor,omitempty"`
	Parent      string `yaml:"parent" json:"parent,omitempty"`             // 父通道引用，格式 provider/service/channel
	ChannelName string `yaml:"channel_name" json:"channel_name,omitempty"` // Channel 显示名称（可选，未配置时回退到 channel）
	ListedSince string `yaml:"listed_since,omitempty" json:"listed_since"` // 收录日期（可选，格式 "2006-01-02"），用于计算收录天数
	ExpiresAt   string `yaml:"expires_at" json:"expires_at,omitempty"`     // 到期日期（可选，格式 "2006-01-02"），过期后赞助等级降为 pulse（板块由可用率决定，不随到期变动）
	// Template 引用 templates/<name>.json 模板，定义完整的请求方式（url/method/headers/body/response）
	Template string `yaml:"template" json:"template,omitempty"`

	// BaseURL 服务商基础地址（如 "https://api.88code.com"），模板通过 {{BASE_URL}} 引用
	BaseURL string `yaml:"base_url,omitempty" json:"base_url"`

	// SkipURLValidation 跳过该监测项的私网 IP 告警校验（内网/自建探测目标场景）
	SkipURLValidation bool `yaml:"skip_url_validation,omitempty" json:"-"`

	// UserIDRefreshMinutes user_id 刷新间隔（分钟），0 = 使用确定性固定值
	UserIDRefreshMinutes int `yaml:"user_id_refresh_minutes" json:"user_id_refresh_minutes,omitempty"`

	// URLPattern URL 模式（含 {{BASE_URL}} 等占位符）
	// 通常由 ResolveTemplates 从模板填充；也可在 config 中显式设置以覆盖模板
	// 在探测期通过 InjectVariables 替换为最终 URL
	URLPattern string `yaml:"url_pattern,omitempty" json:"-"`

	Method  string            `yaml:"method,omitempty" json:"method"`
	Headers map[string]string `yaml:"headers,omitempty" json:"headers"`
	Body    string            `yaml:"body,omitempty" json:"body"`

	// SuccessContains 可选：响应体需包含的关键字，用于判定请求语义是否成功
	SuccessContains string `yaml:"success_contains,omitempty" json:"success_contains"`

	// EnvVarName 可选：自定义环境变量名（用于解决channel名称冲突）
	// 如果指定，则使用此名称覆盖 APIKey，否则使用自动生成的 MONITOR_{PROVIDER}_{SERVICE}_{CHANNEL}_API_KEY
	EnvVarName string `yaml:"env_var_name,omitempty" json:"-"`

	// 自定义巡检间隔（可选，留空则使用全局 interval）
	// 支持 Go duration 格式，例如 "30s"、"1m"、"5m"
	// 商务/赞助等级可按需配置更短间隔
	Interval string `yaml:"interval,omitempty" json:"interval"`

	// 彻底停用配置：不探测、不存储、不展示
	// Disabled 为 true 时，调度器不会创建任务，API 不返回，探测结果不写库
	Disabled       bool   `yaml:"disabled,omitempty" json:"disabled"`
	DisabledReason string `yaml:"disabled_reason,omitempty" json:"disabled_reason"` // 停用原因（可选）

	// 临时下架配置：隐藏但继续探测，用于商家整改期间
	// Hidden 为 true 时，API 不返回该监测项，但调度器继续探测并存储结果
	Hidden       bool   `yaml:"hidden,omitempty" json:"hidden"`
	HiddenReason string `yaml:"hidden_reason,omitempty" json:"hidden_reason"` // 下架原因（可选）

	// 热板/备板/冷板配置：冷板项停止探测，仅展示历史数据（需 boards.enabled=true）
	// Board 可选值：空/"hot"（默认主板）、"secondary"（备板）、"cold"（冷板）。
	// 配置板位是自动移板的"锚点/天花板"：board=secondary 不会被自动升 hot（详见 BoardAutoMoveConfig）。
	Board      string `yaml:"board,omitempty" json:"board"`
	ColdReason string `yaml:"cold_reason" json:"cold_reason,omitempty"` // 冷板原因（可选）

	// 质量移板运行时字段（yaml:"-" 不持久化到 monitors.d，由 automove 在运行时注入）。
	BoardReason       string `yaml:"-" json:"board_reason,omitempty"`        // 移板机器码（如 "quality_hardfail"），前端本地化；运行时注入
	BoardReasonModels string `yaml:"-" json:"board_reason_models,omitempty"` // 触发质量移板的模型名（逗号连接）；运行时注入

	// AutoColdExempt 手动解除自动冷板。
	// 设为 true 时会清除 runtime cold override，并在保持为 true 期间不再自动移入冷板。
	AutoColdExempt bool `yaml:"auto_cold_exempt" json:"auto_cold_exempt,omitempty"`

	// AutoMoveExempt 豁免所有基于可用率的自动移板（hot↔secondary↔cold）。
	// 不影响 expires_at 到期降级逻辑。
	// 设为 true 后，已有的 availability-based override 也会被立即清除。
	AutoMoveExempt bool `yaml:"auto_move_exempt" json:"auto_move_exempt,omitempty"`

	// 通道级慢请求阈值（可选，覆盖模板和全局 slow_latency）
	// 支持 Go duration 格式，例如 "5s"、"15s"
	// 优先级：monitor > template.probe > global
	SlowLatency string `yaml:"slow_latency,omitempty" json:"slow_latency"`

	// 解析后的"慢请求"阈值，用于黄灯判定
	SlowLatencyDuration time.Duration `yaml:"-" json:"-"`

	// 通道级超时时间（可选，覆盖模板和全局 timeout）
	// 支持 Go duration 格式，例如 "10s"、"30s"
	// 优先级：monitor > template.probe > global
	Timeout string `yaml:"timeout,omitempty" json:"timeout"`

	// 解析后的超时时间
	TimeoutDuration time.Duration `yaml:"-" json:"-"`

	// 通道级重试次数（可选，覆盖模板和全局 retry）
	// 0 表示不重试；该字段表示"额外重试次数"，不包含首次尝试
	// 使用 *int 以区分"未设置(nil)"和"显式设置为 0"
	// 优先级：monitor > template.probe > global
	Retry *int `yaml:"retry" json:"retry,omitempty"`

	// 解析后的重试次数（内部使用）
	RetryCount int `yaml:"-" json:"-"`

	// 通道级退避基准间隔（可选，覆盖模板和全局 retry_base_delay）
	// 支持 Go duration 格式，例如 "200ms"、"500ms"
	RetryBaseDelay string `yaml:"retry_base_delay" json:"retry_base_delay,omitempty"`

	// 解析后的退避基准间隔（内部使用）
	RetryBaseDelayDuration time.Duration `yaml:"-" json:"-"`

	// 通道级退避最大间隔（可选，覆盖模板和全局 retry_max_delay）
	// 支持 Go duration 格式，例如 "2s"、"5s"
	RetryMaxDelay string `yaml:"retry_max_delay" json:"retry_max_delay,omitempty"`

	// 解析后的退避最大间隔（内部使用）
	RetryMaxDelayDuration time.Duration `yaml:"-" json:"-"`

	// 通道级抖动比例（可选，覆盖模板和全局 retry_jitter）
	// 取值范围 0-1，0 表示无抖动
	// 使用 *float64 以区分"未设置(nil)"和"显式设置为 0"
	RetryJitter *float64 `yaml:"retry_jitter" json:"retry_jitter,omitempty"`

	// 解析后的抖动比例（内部使用）
	RetryJitterValue float64 `yaml:"-" json:"-"`

	// 解析后的巡检间隔（可选，为空时使用全局 interval）
	IntervalDuration time.Duration `yaml:"-" json:"-"`

	// Proxy 可选：该监测项使用的代理地址
	// 支持格式：
	//   - HTTP/HTTPS 代理：http://host:port, http://user:pass@host:port
	//   - SOCKS5 代理：socks5://host:port, socks5://user:pass@host:port
	//   - socks:// 是 socks5:// 的别名
	// 注意：SOCKS5 代理必须指定端口
	// 不配置时使用系统环境变量代理（HTTP_PROXY/HTTPS_PROXY）
	Proxy string `yaml:"proxy" json:"proxy,omitempty"` // admin monitor CRUD 需要 round-trip；公共 API 不直接序列化 ServiceConfig

	APIKey string `yaml:"api_key" json:"api_key,omitempty"` // admin monitor CRUD 需要 round-trip；公共 API 不直接序列化 ServiceConfig
}

// disabledProviderConfig 批量禁用指定 provider 的配置
// 用于彻底停用某个服务商的所有监测项（不探测、不存储、不展示）
type disabledProviderConfig struct {
	Provider string `yaml:"provider" json:"provider"` // provider 名称，需与 monitors 中的 provider 完全匹配
	Reason   string `yaml:"reason" json:"reason"`     // 停用原因（可选）
}

// hiddenProviderConfig 批量隐藏指定 provider 的配置
// 用于临时下架某个服务商的所有监测项
type hiddenProviderConfig struct {
	Provider string `yaml:"provider" json:"provider"` // provider 名称，需与 monitors 中的 provider 完全匹配
	Reason   string `yaml:"reason" json:"reason"`     // 下架原因（可选）
}

// channelDetailsProviderConfig provider 级通道技术细节暴露配置
// 用于针对特定 provider 覆盖全局 expose_channel_details 设置
type channelDetailsProviderConfig struct {
	Provider string `yaml:"provider" json:"provider"` // provider 名称，匹配时忽略大小写和首尾空格
	Expose   bool   `yaml:"expose" json:"expose"`     // 是否暴露该 provider 的通道技术细节
}

File: frontend/src/types/monitor.ts
/** monitors.d/ 文件元数据 */
export interface MonitorFileMeta {
  source: string;
  revision: number;
  created_at: string;
  updated_at: string;
}

/** 列表页活化用的最新探测快照 */
export interface LatestProbeSnapshot {
  status: number; // 1=绿 2=黄 0=红
  sub_status?: string;
  http_code?: number;
  latency: number; // ms
  timestamp: number; // Unix 秒
  model?: string;
}

/** 监测项摘要（列表用） */
export interface MonitorSummary {
  key: string;
  provider: string;
  service: string;
  channel: string;
  channel_name?: string;
  model_count: number;
  disabled: boolean;
  hidden: boolean;
  board: string;
  category: string;
  template: string;
  source: string;
  revision: number;
  updated_at: string;
  /** 最近探测快照，nil 表示该通道还没探测记录（新建或长期归档） */
  latest_probe?: LatestProbeSnapshot;
}

/** ServiceConfig 的前端子集（详情/编辑用） */
export interface MonitorConfig {
  provider: string;
  provider_name?: string;
  provider_slug?: string;
  provider_url?: string;
  service: string;
  service_name?: string;
  channel: string;
  channel_name?: string;
  model?: string;
  /**
   * monitors.d 行的稳定身份（`md_<uuid>`），由后端生成、不可变。热力图历史按它查，
   * 所以它一变就等于断历史——**编辑时必须原样带回**，别在 payload 里丢掉。
   */
  model_id?: string;
  /** 模型厂商受控 code（internal/modelvendor）；native 族模板必须行级填，否则厂商列显示未知 */
  model_vendor?: string;
  parent?: string;
  template?: string;
  base_url?: string;
  api_key?: string;
  proxy?: string;
  method?: string;
  headers?: Record<string, string>;
  body?: string;
  success_contains?: string;
  category?: string;
  sponsor?: string;
  sponsor_url?: string;
  sponsor_level?: string;
  key_type?: string;
  auto_cold_exempt?: boolean;
  auto_move_exempt?: boolean;
  board?: string;
  cold_reason?: string;
  board_reason?: string;
  board_reason_models?: string;
  retry?: number | null;
  retry_base_delay?: string;
  retry_max_delay?: string;
  retry_jitter?: number | null;
  user_id_refresh_minutes?: number;
  disabled?: boolean;
  disabled_reason?: string;
  hidden?: boolean;
  hidden_reason?: string;
  interval?: string;
  slow_latency?: string;
  timeout?: string;
  listed_since?: string;
  expires_at?: string;
  price_min?: number | null;
  price_max?: number | null;
}

/** monitors.d/ 文件完整结构 */
export interface MonitorFile {
  metadata: MonitorFileMeta;
  monitors: MonitorConfig[];
}

/** 某通道文件下的可探测目标（父或子通道）。
 *  model 取自后端 runtime 已解析配置，是探测请求 target_model 的稳定标识；
 *  未热重载时可能为空（前端据此禁用该行测试按钮）。 */
export interface ProbeTarget {
  role: 'parent' | 'child';
  /** 探测选择器（runtime 已解析的展示名），发给 `/probe` 的 target_model 用它。 */
  model: string;
  /** 停用/启用选择器（monitors.d 行身份）。展示名对模板驱动子行是空串且同父下不唯一，只有它能定位到行。 */
  model_id?: string;
  template: string;
  /**
   * **有效** disabled（父子继承后），不是磁盘值：父通道停用时子行恒为 true。
   * 渲染「停用/启用」文案必须回读 `monitorFile.monitors` 里的原始行，别用这个。
   */
  disabled: boolean;
}

/**
 * 切换某一行的 disabled/hidden。
 *
 * 刻意做成判别联合而不是「可选 modelId」：漏传目标行会让请求退化成「切父通道」，
 * 即用户以为停一个子模型、实际停掉整条通道。有了 `scope` 标签，`{scope:'model'}`
 * 少写 modelId 直接编译不过——把这个事故挡在 tsc 而不是运行时。
 *
 * `scope` 只存在于前端，不上 wire：后端按「model_id 缺省 = 父行」保持向后兼容，
 * 并对显式空值/null 回 400。
 */
export type MonitorToggleRequest =
  | { scope: 'channel'; field: 'disabled' | 'hidden'; value: boolean }
  | { scope: 'model'; field: 'disabled' | 'hidden'; value: boolean; modelId: string };

/** Admin Monitor API 响应 */
export interface AdminMonitorListResponse {
  monitors: MonitorSummary[];
  total: number;
}

export interface AdminMonitorDetailResponse {
  monitor: MonitorFile;
  probe_targets?: ProbeTarget[];
}

/** 单条探测历史记录（管理后台 logs tab 用） */
export interface ProbeHistoryEntry {
  id: number;
  provider: string;
  service: string;
  channel: string;
  model?: string;
  status: number; // 1=绿 2=黄 0=红
  sub_status: string;
  http_code: number;
  latency: number; // ms
  timestamp: number; // Unix 秒
  error_detail?: string;
}

/** 管理后台日志查询响应 */
export interface AdminMonitorLogsResponse {
  logs: ProbeHistoryEntry[];
  total: number;
}

/** rpdiag 质量分 sparkline 数据。
 *  ranking-export.v5.2 起额外暴露 recent_scores（最近 ≤3 次单 sample 升序），
 *  消费方可叠在 30d / 7d 均值之后构造 5 点 sparkline；旧版 wire 仍走 3 点
 *  (avg_30d → avg_7d → latest) fallback。 */
export interface RpdiagScoreTrend {
  latest?: number | null;
  latest_at?: string | null;
  avg_7d?: number | null;
  avg_30d?: number | null;
  /** v5.2+: 最近 ≤3 个 sample 升序（旧→新）。null/缺失时走 latest fallback。 */
  recent_scores?: number[] | null;
  /** v5.4+: 最近 ≤3 次质量相关 terminal attempt 升序（旧→新）。number=打分样本，
   *  null=hard-fail（画灰点）。存在时 sparkline slot 2/3/4 改用它；缺失走 recent_scores。 */
  recent_attempts?: (number | null)[] | null;
  n_7d: number;
  n_30d: number;
}

/** rpdiag (provider, service, channel) 三元组下某个 model 的质量分。 */
export interface RpdiagModelScore {
  model?: string;
  model_key?: string;
  score?: number | null;
  trend: RpdiagScoreTrend;
  detail_url?: string;
  /** rpdiag 判定该 model 当前处于硬失败故障态：score/trend 已被后端归一化为 0
   *  （sparkline 最右点红、贴底）。仅作信息标记，渲染由归一化后的 trend 驱动。 */
  failed?: boolean;
  /** 故障原因文案（rpdiag 后端动态生成），在质量列 tooltip 内 verbatim 展示。 */
  availability_warning?: string;
  /** rpdiag quality_state="unavailable" 且非 hard-fail-active（v5.10 stale/aged 行）。
   *  前端画中性灰「不可测」但保留 recent_attempts 历史彩点（区别于 failed 的清零灰）。 */
  unavailable?: boolean;
  /** rpdiag attempts_7d==0（近 7 天无终态评测记录）且非 hard-fail。展示层把该 model
   *  的真实历史 sparkline 降饱和 + tooltip 注「近7天无评测记录」，与 failed/unavailable
   *  的清零灰不同——这里保留真实历史颜色、只是"发暗"。 */
  no_recent_attempts?: boolean;
}

/** rpdiag 一个 (provider, service, channel) 三元组的聚合质量分。
 *  max_score 是历史 wire 字段名，现为通道内「全站活跃 model」排名贡献的
 *  均值：fresh 行贡献 30 天均值（avg_30d），hard-fail/stale 行计 0，
 *  全站退役 model 从分子分母整体剔除；无活跃 model 时为 null（排序沉底）。
 */
export interface RpdiagScore {
  max_score?: number | null;
  models: RpdiagModelScore[];
  trend: RpdiagScoreTrend;
  channel_url: string;
}

/** /api/rpdiag-scores 响应。
 *  键格式 "provider|service|channel"，三段均小写；channel 为原始 channel_name
 *  （不剥前缀，o-cx/u-cx 等保持区分，见 useRpdiagScores.buildRpdiagKey）。 */
export type RpdiagScoresResponse = Record<string, RpdiagScore>;
